crates/doc/src/shape/widen.rs

```diff
@@ -2,232 +2,393 @@
+            self.minimum = Some(min);
+            self.maximum = Some(max);
+            changed = true;
+        } else {
+            if let Some(min) = &mut self.minimum {
+                if *min > num {
+                    *min = number_bounds(num).0;
+                    changed = true;
+                }
             }
-            crate::Node::Bool(_) => {
-                self.type_ = self.type_ | types::BOOLEAN;
-                false
+            if let Some(max) = &mut self.maximum {
+                if *max < num {
+                    *max = number_bounds(num).1;
+                    changed = true;
+                }
             }
-            crate::Node::Bytes(_) => {
-                self.type_ = self.type_ | types::STRING;
+        }
 
-                let partial_stringshape = StringShape {
-                    content_encoding: Some("base64".to_string()),
-                    ..StringShape::new()
-                };
+        changed
+    }
+}
 
-                self.string = StringShape::union(self.string.clone(), partial_stringshape);
-                false
-            }
-            crate::Node::Null => {
-                self.type_ = self.type_ | types::NULL;
-                false
+impl Shape {
+    pub fn widen<'n, N>(&mut self, node: &'n N) -> bool
+    where
+        N: AsNode,
+    {
+        use json::Number;
+
+        if let Some(_) = &self.enum_ {
+            return self.widen_enum(node);
+        }
+
+        let mut apply_type = |type_| -> bool {
+            let is_first = self.type_ & type_ == types::INVALID;
+            if is_first {
+                self.type_ = self.type_ | type_;
             }
-            crate::Node::Number(num) => {
-                self.type_ = self.type_ | types::Set::for_number(&num);
-                false
+            is_first
+        };
+
+        match node.as_node() {
+            Node::Array(items) => {
+                let is_first = apply_type(types::ARRAY);
+                if is_first {
+                    self.array.additional_items = Some(Box::new(Shape::nothing()));
+                }
+                self.array.widen(items, is_first)
             }
-            crate::Node::String(s) => {
-                let is_first_time = !self.type_.overlaps(types::STRING);
-
-                self.type_ = self.type_ | types::STRING;
-
-                // Similar to the nuance around `is_required`, string format
-                // should only "become detected" the very first time. We still
-                // need to run `detect_format` on subsequent strings because
-                // `StringShape::union()` can sometimes widen a string format,
-                // e.g from `integer` -> `number`
-                let format = if is_first_time || self.string.format.is_some() {
-                    Format::detect(s)
+            Node::Bool(_) => apply_type(types::BOOLEAN),
+            Node::Bytes(_) => panic!("not implemented"),
```

Comment:
This was implemented previously, as a string with `content_encoding: "base64"`. (why) do we not care about this case?

Reply:
Node::Bytes isn't used today -- it exists somewhat speculatively, as a way to speed up not-yet-implemented reduction annotations that _may_ work over bytes.

Potentially it ought to be removed, except for a gut feel that it will be important one day. So, panic because it can never actually be encountered.